File: Graph Algorithms/Shortest_Routes_II.py
import sys,math,random
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations

# functions #
MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(2**62)
def gcd(a,b):
    if a%b==0:
        return b
    else:
        return gcd(b,a%b)
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
##

#String hashing : sh, fenwick sortedlist : fsortl, Number : numtheory, SparseTable : SparseTable
#bucket sorted list : bsortl, segment tree(lazy propogation) : SegmentTree, bootstrap : bootstrap
#binary indexed tree : BIT, segment tree(point updates) : SegmentPoint, Convex Hull : hull
#Combinatorics : pnc, Diophantine Equations : dpheq
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
 
def dijkstra(graph, start,n):
    """ 
        Uses Dijkstra's algortihm to find the shortest path from node start
        to all other nodes in a directed weighted graph.
    """
    dist, parents = [float("inf")] * n, [-1] * n
    dist[start] = 0
 
    queue = [(0, start)]
    while queue:
        path_len, v = heappop(queue)
        if path_len == dist[v]:
            for w, edge_len in graph[v]:
                if edge_len + path_len < dist[w]:
                    dist[w], parents[w] = edge_len + path_len, v
                    heappush(queue, (edge_len + path_len, w))
    return dist
 
def solve():
    L = list(map(int, sys.stdin.readline().split()))
    d =defaultdict(list)
    for i in range(L[1]):
        L1 = list(map(int, sys.stdin.readline().split()))
        d[L1[0]].append([L1[1],L1[2]])
        d[L1[1]].append([L1[0],L1[2]])
    
    d1 = defaultdict(list)
 
    for i in range(L[2]):
        L1 = list(map(int, sys.stdin.readline().split()))
        L1.sort()
        if not d1[L1[0]]:
            d1[L1[0]] = dijkstra(d,L1[0],L[0]+1)
        k = d1[L1[0]][L1[1]]
        if k==float('inf'):
            print(-1)
            continue
        print(k)
    #st = sys.stdin.readline().strip()

    #st = sys.stdin.readline().strip()
solve()

File: Graph Algorithms/Shortest_Routes_II.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
using namespace std;
#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define print(arr) for (auto it : arr){cout<<it<<" ";}cout<<endl;
#define len(arr) arr.size()
#define printf(x) cout << x << endl;
#define printm(map) cout<<"{";for(auto it: map){cout<<it.first<<":"<<it.second<<",";};cout<<"}"<<endl;

// const int MOD = 998244353;
const int MOD = 1e9 + 7;

int gcd(int a, int b) {
    if (b == 0) return a;
    return gcd(b, a % b);
}

int lcm(int a, int b) {
    return a / gcd(a, b) * b;
}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie/Treap: Tries
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
// Persistent Segment Tree: perseg, FreqGraphs: bgraph
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

void solve() {
    int n,m,q;
    cin>>n>>m>>q;

    int dist[n+1][n+1];
    
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            dist[i+1][j+1] = INF;
        }
        dist[i+1][i+1] = 0;
    }

    while(m--){
        int a,b,c;
        cin>>a>>b>>c;
        dist[a][b] = min(dist[a][b],c);
        dist[b][a] = min(dist[b][a],c);
    }
    // printf(dist[1][4])

    for(int i=1;i<=n;i++){
        for(int j=1;j<=n;j++){
            for(int k=1;k<=n;k++){
                if(dist[i][k]!=INF && dist[j][i]!=INF){
                    dist[j][k] = min(dist[j][k],dist[j][i]+dist[i][k]);
                }
            }
        }
    }
    // printf(dist[1][4])

    while(q--){
        int a,b;
        cin>>a>>b;
        printf((((dist[a][b])!=INF)?dist[a][b]:-1))
    }
}

int32_t main() {
    fastio
    int t=1;
    while (t--) solve();
    return 0;
}